pallets/system/src/header_builder.rs

```diff
@@ -263,3 +264,207 @@ impl Decode for AvailExtrinsic {
 		})
 	}
 }
+
+fn data_root_avail(app_ext: Vec<AppExtrinsic>) -> [u8; 32] {
+	let ext: Vec<Vec<u8>> = app_ext.iter().map(|x| x.data.clone()).collect();
+	let avail_ext = ext
+		.iter()
+		.filter_map(|e| AvailExtrinsic::decode(&mut &e[..]).ok())
+		.collect::<Vec<_>>();
+	let data_root: [u8; 32] = if avail_ext.len() > 0 {
+		let leaves: Vec<[u8; 32]> = avail_ext.iter().map(|e| Sha256::hash(&e.data)).collect();
+		let tree = MerkleTree::<Sha256>::from_leaves(&leaves);
+		tree.root().expect("Data Root computation failed")
+	} else {
+		Default::default()
+	};
+	data_root
+}
+
+mod tests {
+	use super::*;
+
+	/// tests for data root is created correctly
+	#[test]
+	fn data_root_test() {
```

Comment:
Probably we can use hex and test_case macros to enumerate cases and make them clearer and shorter.